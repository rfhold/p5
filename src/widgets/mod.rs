mod resource_list;
mod scrollable_list;
mod splash;
mod stack_config;
mod stack_layout;
mod stack_operation;
mod stack_outputs;
mod stack_resources;

pub use resource_list::ResourceListState;
pub use splash::*;
pub use stack_layout::*;
