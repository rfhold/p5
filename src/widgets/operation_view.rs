use ratatui::{
    buffer::Buffer,
    layout::{Alignment, Rect},
    widgets::{Block, Paragraph, StatefulWidget, Widget},
};

use crate::{AppState, state::Loadable};

#[derive(Clone, Default)]
pub struct OperationView {}

impl StatefulWidget for OperationView {
    type State = AppState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let workspace = state.workspace();

        let (title, message) = match workspace {
            Loadable::Loaded(workspace) => match state.stack() {
                Loadable::Loaded(stack) => {
                    let title = format!("Workspace: {} - Stack: {}", workspace.cwd, stack.name);

                    match state.stack_outputs() {
                        Loadable::Loaded(outputs) => (
                            title,
                            serde_json::to_string_pretty(&outputs)
                                .unwrap_or_else(|_| "Failed to serialize outputs".to_string()),
                        ),
                        Loadable::Loading => (title, "Loading Outputs...".to_string()),
                        Loadable::NotLoaded => (title, "No Outputs Available".to_string()),
                    }
                }
                Loadable::Loading => (
                    format!("Workspace: {}", workspace.cwd),
                    "Loading Stack...".to_string(),
                ),
                Loadable::NotLoaded => (
                    format!("Workspace: {}", workspace.cwd),
                    "No Stack Selected".to_string(),
                ),
            },
            Loadable::Loading => (
                "Loading Workspace...".to_string(),
                "Please wait...".to_string(),
            ),
            Loadable::NotLoaded => (
                "No Workspace Loaded".to_string(),
                "Please load a workspace.".to_string(),
            ),
        };
        let block = Block::bordered()
            .title(title)
            .border_type(ratatui::widgets::BorderType::Rounded);

        let paragraph = Paragraph::new(message)
            .block(block)
            .alignment(Alignment::Left);

        paragraph.render(area, buf);
    }
}
