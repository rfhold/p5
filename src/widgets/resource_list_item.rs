use pulumi_automation::stack::{Operation, ResourceState, ResourceType, StackChangeStep};
use ratatui::{
    style::{Color, Style},
    text::{Line, Span},
    widgets::ListItem,
};

pub struct ResourceListItem<'resource> {
    pub operation: Option<Operation>,
    pub resource_state: Option<&'resource ResourceState>,
}

impl<'a> Into<ListItem<'a>> for ResourceListItem<'a> {
    fn into(self) -> ListItem<'a> {
        if let Some(resource) = self.resource_state {
            let mut line = Line::default();

            if let Some(op) = self.operation {
                line.push_span(Span::styled(
                    op.to_string(),
                    Style::default().fg(match op {
                        Operation::Create => Color::LightGreen,
                        Operation::Update => Color::LightYellow,
                        Operation::Delete => Color::LightRed,
                        Operation::Same => Color::DarkGray,
                        _ => Color::LightBlue,
                    }),
                ));

                line.push_span(" ");
            }

            line.push_span(Span::styled(
                resource.resource_type.to_string(),
                Style::default().fg(match resource.resource_type {
                    ResourceType::Stack => Color::DarkGray,
                    _ => Color::Gray,
                }),
            ));

            line.push_span(" ");

            line.push_span(resource.name().to_string());

            ListItem::new(line)
        } else {
            ListItem::new(Line::default())
        }
    }
}

impl<'a> From<&'a ResourceState> for ResourceListItem<'a> {
    fn from(resource_state: &'a ResourceState) -> Self {
        Self {
            resource_state: Some(resource_state),
            operation: None,
        }
    }
}

impl<'a> From<&'a StackChangeStep> for ResourceListItem<'a> {
    fn from(step: &'a StackChangeStep) -> Self {
        if step.new_state.is_none() {
            return Self {
                resource_state: step.old_state.as_ref(),
                operation: Some(step.op.clone()),
            };
        }
        Self {
            resource_state: step.new_state.as_ref(),
            operation: Some(step.op.clone()),
        }
    }
}
