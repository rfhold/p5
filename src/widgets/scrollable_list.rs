use ratatui::{
    buffer::Buffer,
    layout::{Margin, Rect},
    style::Style,
    widgets::{
        Block, List, ListItem, ListState, Scrollbar, ScrollbarOrientation, ScrollbarState,
        StatefulWidget,
    },
};

#[derive(Default, Clone, Debug)]
pub struct ScrollableListState {
    pub list_state: ListState,
}

pub struct ScrollableList<'a> {
    items: Vec<ListItem<'a>>,
    block: Option<Block<'a>>,
}

impl<'a> ScrollableList<'a> {
    pub fn new(items: Vec<ListItem<'a>>) -> Self {
        Self { items, block: None }
    }

    pub fn block(mut self, block: Block<'a>) -> Self {
        self.block = Some(block);
        self
    }

    fn content_height(&self) -> usize {
        self.items.iter().map(|item| item.height()).sum::<usize>()
    }

    fn v_scroll_state(&self, list_state: &ListState, area: &Rect) -> Option<ScrollbarState> {
        let content_height = self.content_height();
        let area_height = area.height as usize;
        if content_height <= area_height {
            return None;
        }
        let position = self
            .items
            .iter()
            .take(list_state.selected().unwrap_or(0))
            .map(|item| item.height())
            .sum::<usize>();

        let overflow = content_height - area_height;
        let offset_position = position.saturating_sub(overflow);

        Some(ScrollbarState::new(overflow).position(offset_position))
    }

    fn v_scroll_area(&self, area: Rect) -> Rect {
        if let Some(_block) = &self.block {
            area.inner(Margin {
                horizontal: 0,
                vertical: 1,
            })
        } else {
            area
        }
    }
}

impl StatefulWidget for ScrollableList<'_> {
    type State = ScrollableListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let v_scroll_area = self.v_scroll_area(area);
        let v_scroll_state = self.v_scroll_state(&state.list_state, &v_scroll_area);

        let list = List::new(self.items)
            .block(self.block.unwrap_or_default())
            .highlight_style(Style::default().fg(ratatui::style::Color::Yellow));

        StatefulWidget::render(list, area, buf, &mut state.list_state);

        if v_scroll_state.is_none() {
            return;
        }

        let v_scrollbar = Scrollbar::new(ScrollbarOrientation::VerticalRight);
        StatefulWidget::render(
            v_scrollbar,
            v_scroll_area,
            buf,
            &mut v_scroll_state.unwrap(),
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    #[test]
    fn test_basic_scrollable() {
        let list_items = vec![
            ListItem::new("Item 1"),
            ListItem::new("Item 2"),
            ListItem::new("Item 3"),
            ListItem::new("Item 4"),
        ];

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_overflow() {
        let list_items = vec![
            ListItem::new("Item 1"),
            ListItem::new("Item 2"),
            ListItem::new("Item 3"),
            ListItem::new(
                "Item 4, which is a bit longer than the others to test overflow. Turns out, it still needs to be a bit longer",
            ),
        ];

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(30)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll_first_page() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(17)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll_last_page() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(34)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
