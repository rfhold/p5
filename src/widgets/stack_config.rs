use ratatui::{
    layout::Alignment,
    widgets::{Block, Paragraph, StatefulWidget, Widget},
};

use crate::state::{AppState, Loadable};

#[derive(Default, Clone)]
pub struct StackConfig {}

impl StatefulWidget for StackConfig {
    type State = AppState;

    fn render(
        self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
    ) {
        let config = match state.stack_config() {
            Loadable::Loaded(config) => Some(
                serde_json::to_string_pretty(&config)
                    .unwrap_or_else(|_| "Failed to serialize config".to_string()),
            ),
            Loadable::Loading => Some("Loading Config...".to_string()),
            Loadable::NotLoaded => None,
        };

        if let Some(config) = config {
            let config_block = Block::bordered()
                .title("Config")
                .border_type(ratatui::widgets::BorderType::Rounded);

            let config_paragraph = Paragraph::new(config)
                .block(config_block)
                .alignment(Alignment::Left);

            config_paragraph.render(area, buf);
        }
    }
}
