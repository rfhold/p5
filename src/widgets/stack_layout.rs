use ratatui::{
    layout::{Alignment, Constraint, Direction, Layout},
    widgets::{Block, Paragraph, StatefulWidget, Widget},
};

use crate::state::{AppState, Loadable, StackContext};

use super::{
    stack_config::StackConfig, stack_outputs::StackOutputs, stack_preview::StackPreview,
    stack_resources::StackResources, stack_update::StackUpdate,
};

pub struct StackLayout {
    pub context: StackContext,
}

impl StackLayout {
    pub fn new(context: StackContext) -> Self {
        Self { context }
    }
}

impl StatefulWidget for StackLayout {
    type State = AppState;

    fn render(
        self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
    ) {
        let workspace = state.workspace();
        let stack = state.stack();

        let title = match workspace {
            Loadable::Loaded(workspace) => format!("Workspace: {}", workspace.cwd),
            Loadable::Loading => "Loading Workspace...".to_string(),
            Loadable::NotLoaded => "No Workspace Loaded".to_string(),
        };

        let body = match stack {
            Loadable::Loaded(stack) => format!("Stack: {}", stack.name),
            Loadable::Loading => "Loading Stack...".to_string(),
            Loadable::NotLoaded => "No Stack Selected".to_string(),
        };

        let title_block = Block::bordered()
            .title(title)
            .border_type(ratatui::widgets::BorderType::Rounded);

        let body_paragraph = Paragraph::new(body)
            .block(title_block.clone())
            .alignment(Alignment::Left);

        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints(vec![Constraint::Length(3), Constraint::Percentage(90)])
            .split(area);

        body_paragraph.render(layout[0], buf);

        match self.context {
            StackContext::Outputs => StackOutputs::default().render(layout[1], buf, state),
            StackContext::Config => StackConfig::default().render(layout[1], buf, state),
            StackContext::Resources => StackResources::default().render(layout[1], buf, state),
            StackContext::Preview => StackPreview::default().render(layout[1], buf, state),
            StackContext::Update => StackUpdate::default().render(layout[1], buf, state),
        }
    }
}
