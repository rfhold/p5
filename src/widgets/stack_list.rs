use ratatui::{
    layout::Alignment,
    widgets::{Block, List, ListItem, Paragraph, StatefulWidget, Widget},
};

use crate::state::{AppState, Loadable};

pub struct StackList {}

impl StackList {
    pub fn new() -> Self {
        StackList {}
    }
}

impl StatefulWidget for StackList {
    type State = AppState;

    fn render(
        self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
    ) {
        let block = Block::bordered()
            .title("Stacks")
            .border_type(ratatui::widgets::BorderType::Rounded);

        match state.stacks() {
            Loadable::Loaded(stacks) => {
                let items = stacks
                    .iter()
                    .map(|s| ListItem::new(s.name.clone()))
                    .collect::<Vec<_>>();

                let list = List::new(items).block(block);
                StatefulWidget::render(list, area, buf, &mut Default::default());
            }
            Loadable::Loading => {
                Paragraph::new("Loading Stacks...".to_string())
                    .block(block)
                    .alignment(Alignment::Left)
                    .render(area, buf);
            }
            Loadable::NotLoaded => {
                Paragraph::new("No Workspace Selected".to_string())
                    .block(block)
                    .alignment(Alignment::Left)
                    .render(area, buf);
            }
        };
    }
}
