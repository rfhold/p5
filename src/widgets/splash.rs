use ratatui::widgets::Widget;

pub struct SplashScreen {
    pub title: String,
    pub message: String,
}

impl SplashScreen {
    pub fn new(title: String, message: String) -> Self {
        Self { title, message }
    }
}

impl Widget for SplashScreen {
    fn render(self, area: ratatui::prelude::Rect, buf: &mut ratatui::prelude::Buffer) {
        let title_block = ratatui::widgets::Block::bordered()
            .title(self.title)
            .border_type(ratatui::widgets::BorderType::Rounded);

        let message_paragraph = ratatui::widgets::Paragraph::new(self.message)
            .block(title_block)
            .alignment(ratatui::layout::Alignment::Center);

        message_paragraph.render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    #[test]
    fn test_render_splash_screen() {
        let splash_screen = SplashScreen::new(
            "P5".to_string(),
            "Press ':' to open the command prompt. Ctrl+C to exit.".to_string(),
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(splash_screen, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
