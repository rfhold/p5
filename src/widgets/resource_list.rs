use pulumi_automation::stack::{Operation, ResourceState, StackChangeSummary};
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, ListItem, StatefulWidget},
};
use ratatui_macros::{line, span};

use super::scrollable_list::{ScrollableList, ScrollableListState};
use crate::state::ResourceOperationState;

#[derive(Debug, Clone)]
pub struct ResourceItem<'a> {
    pub urn: String,
    pub line: Line<'a>,
}

pub trait ResourceAdapter {
    fn get_items(&self) -> Vec<ResourceItem>;
}

#[derive(Default, Clone, Debug)]
pub struct ResourceListState {
    pub scrollable_state: ScrollableListState,
    pub excluded_resources: Vec<String>,
    pub target_resources: Vec<String>,
    pub replace_resources: Vec<String>,
}

pub struct ResourceList<'a, A: ResourceAdapter> {
    pub block: Block<'a>,
    pub adapter: A,
}

// Adapter implementations
pub struct CurrentStateAdapter<'a>(pub &'a Vec<ResourceState>);

impl<'a> ResourceAdapter for CurrentStateAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .iter()
            .map(|resource| ResourceItem {
                urn: resource.urn.clone(),
                line: line!(
                    span!(Style::default().fg(Color::White); "{:25} ", resource.resource_type),
                    span!(Style::default().fg(Color::Cyan); "{:<}", resource.name()),
                ),
            })
            .collect()
    }
}

pub struct ChangeSummaryAdapter<'a>(pub &'a StackChangeSummary);

impl<'a> ResourceAdapter for ChangeSummaryAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .steps
            .iter()
            .map(|step| {
                let state = if let Some(s) = step.new_state.as_ref() {
                    Some(s)
                } else {
                    step.old_state.as_ref()
                };

                if let Some(ref state) = state {
                    ResourceItem {
                        urn: state.urn.clone(),
                        line: line!(
                            operation_to_span(&step.op, Some(10)),
                            span!(Style::default().fg(Color::Gray); "{:<25} ", state.resource_type),
                            span!(Style::default(); "{:<}", state.name()),
                        ),
                    }
                } else {
                    ResourceItem {
                        urn: String::new(),
                        line: line!(
                            operation_to_span(&step.op, Some(10)),
                            span!(Style::default().fg(Color::Gray); "{:<25} ", "Unknown resource type"),
                            span!(Style::default(); "{:<}", "Unknown resource"),
                        ),
                    }
                }
            })
            .collect()
    }
}

pub struct OperationStateAdapter<'a>(pub &'a Vec<ResourceOperationState>);

impl<'a> ResourceAdapter for OperationStateAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .iter()
            .map(|op_state| {
                let metadata = match op_state {
                    ResourceOperationState::InProgress { pre_event, .. } => &pre_event.metadata,
                    ResourceOperationState::Completed { out_event, .. } => &out_event.metadata,
                    ResourceOperationState::Failed { failed_event, .. } => &failed_event.metadata,
                };

                let pre_event = match op_state {
                    ResourceOperationState::InProgress { pre_event, .. } => pre_event,
                    ResourceOperationState::Completed { pre_event, .. } => pre_event,
                    ResourceOperationState::Failed { pre_event, .. } => pre_event,
                };

                let (dur_text, dur_style) = {
                    let (start, end) = match op_state {
                        ResourceOperationState::InProgress { start_time, .. } => {
                            (*start_time, chrono::Utc::now())
                        }
                        ResourceOperationState::Completed {
                            start_time,
                            end_time,
                            ..
                        } => (*start_time, *end_time),
                        ResourceOperationState::Failed {
                            start_time,
                            end_time,
                            ..
                        } => (*start_time, *end_time),
                    };

                    let duration_s = (end - start).num_seconds();
                    (
                        format!("{:.0}s", duration_s as f64),
                        Style::default().fg(match op_state {
                            ResourceOperationState::InProgress { .. } => Color::Gray,
                            ResourceOperationState::Completed { .. } => Color::DarkGray,
                            ResourceOperationState::Failed { .. } => Color::Red,
                        }),
                    )
                };

                let resource_type = metadata.resource_type.to_string();

                let name = metadata.name();

                ResourceItem {
                    urn: metadata.urn.clone(),
                    line: line!(
                        operation_to_span(&pre_event.metadata.op, Some(10)),
                        span!(dur_style; "{:<2} ", dur_text),
                        span!(Style::default().fg(Color::Gray); "{:.25} ", resource_type),
                        span!(Style::default(); "{:<}", name),
                    ),
                }
            })
            .collect()
    }
}

impl<'a, A: ResourceAdapter> StatefulWidget for ResourceList<'a, A> {
    type State = ResourceListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let items = self.adapter.get_items();

        let list_items: Vec<ListItem> = items
            .iter()
            .map(|item| {
                let mut style = Style::default();

                // Apply selection styling (priority order matters)
                if state.excluded_resources.contains(&item.urn) {
                    style = style.fg(Color::Red).add_modifier(Modifier::CROSSED_OUT);
                } else if state.replace_resources.contains(&item.urn) {
                    style = style.fg(Color::Yellow).add_modifier(Modifier::ITALIC);
                } else if state.target_resources.contains(&item.urn) {
                    style = style.fg(Color::Green).add_modifier(Modifier::BOLD);
                }

                ListItem::new(item.line.clone()).style(style)
            })
            .collect();

        let scrollable_list = ScrollableList::new(list_items).block(self.block);
        scrollable_list.render(area, buf, &mut state.scrollable_state);
    }
}

impl<'a> ResourceList<'a, CurrentStateAdapter<'a>> {
    pub fn from_states(block: Block<'a>, states: &'a Vec<ResourceState>) -> Self {
        ResourceList {
            block,
            adapter: CurrentStateAdapter(states),
        }
    }
}

impl<'a> ResourceList<'a, ChangeSummaryAdapter<'a>> {
    pub fn from_summary(block: Block<'a>, summary: &'a StackChangeSummary) -> Self {
        ResourceList {
            block,
            adapter: ChangeSummaryAdapter(summary),
        }
    }
}

impl<'a> ResourceList<'a, OperationStateAdapter<'a>> {
    pub fn from_operations(block: Block<'a>, operations: &'a Vec<ResourceOperationState>) -> Self {
        ResourceList {
            block,
            adapter: OperationStateAdapter(operations),
        }
    }
}

fn operation_to_span(op: &Operation, width: Option<usize>) -> Span<'_> {
    let style = match op {
        Operation::Create => Style::default().fg(Color::Green),
        Operation::Update => Style::default().fg(Color::Yellow),
        Operation::Delete => Style::default().fg(Color::Red),
        Operation::Read => Style::default().fg(Color::Blue),
        Operation::Replace => Style::default().fg(Color::Magenta),
        _ => Style::default(),
    };

    match width {
        Some(w) => span!(style; "{:<w$}", op.to_string()),
        None => span!(style; "{}", op.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use pulumi_automation::{
        event::{ResOpFailedDetails, ResOutputsDetails, ResourcePreDetails},
        stack::{ResourceType, StackChangeStep},
    };
    use ratatui::{Terminal, backend::TestBackend, widgets::ListState};

    #[test]
    fn test_resources() {
        let items = vec![
            ResourceState {
                urn: "urn:pulumi:stack::project::pulumi:pulumi:Stack::project".to_string(),
                resource_type: ResourceType::Stack,
                ..Default::default()
            },
            ResourceState {
                urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket".to_string(),
                resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                ..Default::default()
            },
        ];

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_states(
            Block::default()
                .title("Resources")
                .borders(ratatui::widgets::Borders::ALL),
            &items,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_change_summary() {
        let change_summary = StackChangeSummary {
            steps: vec![
                StackChangeStep {
                    op: pulumi_automation::stack::Operation::Create,
                    old_state: None,
                    new_state: Some(ResourceState {
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                StackChangeStep {
                    op: pulumi_automation::stack::Operation::Delete,
                    old_state: Some(ResourceState {
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::old-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        ..Default::default()
                    }),
                    new_state: None,
                    ..Default::default()
                },
            ],
            ..Default::default()
        };

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_summary(
            Block::default()
                .title("Change Summary")
                .borders(ratatui::widgets::Borders::ALL),
            &change_summary,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_operation_states() {
        let operation_states = vec![
            ResourceOperationState::InProgress {
                sequence: 1,
                start_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Create,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
            },
            ResourceOperationState::Completed {
                sequence: 2,
                start_time: chrono::Utc::now(),
                end_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Update,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
                out_event: ResOutputsDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Update,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
            },
            ResourceOperationState::Failed {
                sequence: 3,
                start_time: chrono::Utc::now(),
                end_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Delete,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
                failed_event: ResOpFailedDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Delete,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    steps: 0,
                    status: 0,
                    extra_values: None,
                },
            },
        ];

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_operations(
            Block::default()
                .title("Resource Operations")
                .borders(ratatui::widgets::Borders::ALL),
            &operation_states,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
