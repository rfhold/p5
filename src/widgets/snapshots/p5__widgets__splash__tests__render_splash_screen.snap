---
source: src/widgets/splash.rs
assertion_line: 45
expression: terminal.backend()
---
"╭P5────────────────────────────────────────────────────────────────────────────╮"
"│             Press ':' to open the command prompt. Ctrl+C to exit.            │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"╰──────────────────────────────────────────────────────────────────────────────╯"
